crates/router/src/core/webhooks/network_tokenization_incoming.rs

```diff
@@ -0,0 +1,459 @@
+                .attach_printable("Locker id is not found for the payment method")?,
+            payment_method,
+            merchant_context,
+            decrypted_data,
+            true,
+        )
+        .await
+    }
+}
+
+#[async_trait]
+impl NetworkTokenWebhookResponseExt for pm_types::NetworkTokenMetaDataUpdateBody {
+    fn decrypt_payment_method_data(
+        &self,
+        payment_method: &domain::PaymentMethod,
+    ) -> CustomResult<api::payment_methods::CardDetailFromLocker, errors::ApiErrorResponse> {
+        let decrypted_data = payment_method
+            .network_token_payment_method_data
+            .clone()
+            .map(|x| x.into_inner().expose())
+            .and_then(|v| {
+                serde_json::from_value::<api::payment_methods::PaymentMethodsData>(v).ok()
+            })
+            .and_then(|pmd| match pmd {
+                api::payment_methods::PaymentMethodsData::Card(token) => {
+                    Some(api::payment_methods::CardDetailFromLocker::from(token))
+                }
+                _ => None,
+            })
+            .ok_or(errors::ApiErrorResponse::InternalServerError)
+            .attach_printable("Failed to obtain decrypted token object from db")?;
+        Ok(decrypted_data)
+    }
+
+    async fn update_payment_method(
+        &self,
+        state: &SessionState,
+        payment_method: &domain::PaymentMethod,
+        merchant_context: &domain::MerchantContext,
+    ) -> CustomResult<WebhookResponseTracker, errors::ApiErrorResponse> {
+        let decrypted_data = self.decrypt_payment_method_data(payment_method)?;
+        handle_metadata_update(
+            state,
+            &self.token,
+            payment_method
+                .network_token_locker_id
+                .clone()
+                .get_required_value("locker_id")
+                .change_context(errors::ApiErrorResponse::InternalServerError)
+                .attach_printable("Locker id is not found for the payment method")?,
+            payment_method,
+            merchant_context,
+            decrypted_data,
+            true,
+        )
+        .await
+    }
+}
+
+pub struct Authorization {
+    header: Option<HeaderValue>,
+}
+
+impl Authorization {
+    pub fn new(header: Option<&HeaderValue>) -> Self {
+        Self {
+            header: header.cloned(),
+        }
+    }
+
+    pub async fn verify_webhook_source(
+        self,
+        nt_service: &settings::NetworkTokenizationService,
+    ) -> CustomResult<(), errors::ApiErrorResponse> {
+        let secret = nt_service.webhook_source_verification_key.clone();
+
+        let source_verified = match self.header {
+            Some(authorization_header) => match authorization_header.to_str() {
+                Ok(header_value) => Ok(header_value == secret.expose()),
+                Err(err) => {
+                    logger::error!("Failed to parse authorization header: {}", err);
```

Comment:
You might want to consider using `Debug` impl instead when logging the error.

```suggestion
                    logger::error!("Failed to parse authorization header: {err:?}");
```

Or even:

```suggestion
                    logger::error!(?error, "Failed to parse authorization header");
```